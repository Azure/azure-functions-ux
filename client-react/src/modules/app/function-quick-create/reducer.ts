import { IAction } from '../../../models/action';
// import { cSharpCode } from '../../../pages/app/function-quick-create/components/header-form';
import {
  UPDATE_CURRENT_CODE,
  UPDATE_FUNCTION_APP_LANGUAGE,
  UPDATE_FUNCTION_APP_LANGUAGE_AND_CODE,
  UPDATE_FUNCTION_APP_NAME,
  UPDATE_IS_SUBMITTING,
} from './actions';

export interface IFunctionQuickCreateState {
  // code: string;
  functionAppName: string;
  functionAppLanguage: string;
  isSubmitting: boolean;
}

export const InitialState: IFunctionQuickCreateState = {
  // code: cSharpCode,
  functionAppName: '',
  functionAppLanguage: 'csharp',
  isSubmitting: false,
};

const functionQuickCreate = (state = InitialState, action: IAction<any>) => {
  switch (action.type) {
    case UPDATE_CURRENT_CODE:
      return { ...state, ...action.payload };
    case UPDATE_FUNCTION_APP_NAME:
      return { ...state, ...action.payload };
    case UPDATE_FUNCTION_APP_LANGUAGE:
      return { ...state, ...action.payload };
    case UPDATE_FUNCTION_APP_LANGUAGE_AND_CODE:
      return { ...state, ...action.payload };
    case UPDATE_IS_SUBMITTING:
      return { ...state, ...action.payload };
    default:
      return state;
  }
};

export default functionQuickCreate;
