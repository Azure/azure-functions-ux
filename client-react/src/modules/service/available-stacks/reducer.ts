import { ArmArray } from '../../../models/WebAppModels';
import { IAction } from '../../../models/action';
import { UPDATE_AVAILABLE_STACKS_LOADING, UPDATE_AVAILABLE_STACKS } from './actions';
import { DEFAULT_KEY, generateCacheTTL } from 'redux-cache';
import { AvailableStack } from '../../../models/available-stacks';

export interface IStacksState {
  loading: boolean;
  stacks: ArmArray<AvailableStack>;
}
export const InitialState: IStacksState = {
  [DEFAULT_KEY]: null,
  loading: false,
  stacks: {
    value: [],
  },
};

const stacks = (state = InitialState, action: IAction<any>) => {
  switch (action.type) {
    case UPDATE_AVAILABLE_STACKS_LOADING:
      return { ...state, loading: action.payload };
    case UPDATE_AVAILABLE_STACKS:
      return {
        ...state,
        [DEFAULT_KEY]: generateCacheTTL(60000),
        stacks: action.payload,
      };
    default:
      return state;
  }
};

export default stacks;
