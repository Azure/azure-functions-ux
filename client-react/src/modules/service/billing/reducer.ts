import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import { BillingMeter } from '../../../models/BillingModels';
import { ArmArray } from '../../../models/WebAppModels';
import { metadataReducer } from '../../ApiReducerHelper';
import { ApiState } from '../../types';
import * as actions from './actions';
import { AREA_STRING, BILLING_METERS_FETCH_SUCCESS } from './actionTypes';

export type BillingMetersAction = ActionType<typeof actions>;
export type BillingMetersState = ApiState<ArmArray<BillingMeter>>;

export const InitialState = {
  data: {
    value: [],
  },
};

export default combineReducers<BillingMetersState, BillingMetersAction>({
  metadata: metadataReducer(AREA_STRING),
  data: (state = InitialState.data, action) => {
    switch (action.type) {
      case BILLING_METERS_FETCH_SUCCESS:
        return action.meters;
      default:
        return state;
    }
  },
});
