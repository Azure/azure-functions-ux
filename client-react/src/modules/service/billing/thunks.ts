import { updateBillingMeters, updateBillingMetersLoading } from './actions';
import { checkCacheValid } from 'redux-cache';
import axios from 'axios';
import { ArmArray } from '../../../models/WebAppModels';
import { BillingMeter } from '../../../models/BillingModels';

export type OSType = 'Windows' | 'Linux';
export function fetchBillingMeters(subscriptionId: string, osType: OSType, location?: string) {
  return async (dispatch: any, getState: any) => {
    const startupInfo = getState().portalService.startupInfo;
    const armToken = startupInfo.token;
    const armEndpoint = startupInfo.armEndpoint;
    const currentBillingMetersCall = getState().billingMeters;
    const isCacheValid = checkCacheValid(getState, 'billingMeters');
    if (isCacheValid || currentBillingMetersCall.loading) {
      return;
    }

    dispatch(updateBillingMetersLoading(true));

    try {
      let url = `${armEndpoint}/subscriptions/${subscriptionId}/providers/Microsoft.Web/billingMeters?api-version=2015-08-01`;
      if (location) {
        url += `&billingLocation=${location.replace(/\s/g, '')}`;
      }

      if (osType) {
        url += `&osType=${osType}`;
      }
      const billingMetersFetch = await axios.get<ArmArray<BillingMeter>>(url, {
        headers: {
          Authorization: `Bearer ${armToken}`,
        },
      });
      const billingMetersResult = billingMetersFetch.data;
      dispatch(updateBillingMeters(billingMetersResult));
    } catch (err) {
      console.log(err);
    } finally {
      dispatch(updateBillingMetersLoading(false));
    }
  };
}
