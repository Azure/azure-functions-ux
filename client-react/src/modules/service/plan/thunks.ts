import { updateCurrentPlan, updateCurrentPlanNoCache } from './actions';
import { checkCacheValid } from 'redux-cache';
import axios from 'axios';
import { ArmObj, ServerFarm } from '../../../models/WebAppModels';
import IState from '../../types';

export function fetchPlan() {
  return async (dispatch: any, getState: () => IState) => {
    const startupInfo = getState().portalService.startupInfo;
    if (!startupInfo) {
      return null;
    }
    const armToken = startupInfo.token;
    const armEndpoint = startupInfo.armEndpoint;
    const currentPlan = getState().plan;
    const isCacheValid = checkCacheValid(getState, 'plan');
    if (isCacheValid && currentPlan.resourceId === currentPlan.plan.id && !currentPlan.loading) {
      return currentPlan.plan;
    }

    dispatch(updateCurrentPlanNoCache({ loading: true }));
    try {
      const planFetch = await axios.get<ArmObj<ServerFarm>>(`${armEndpoint}${currentPlan.resourceId}?api-version=2015-08-01`, {
        headers: {
          Authorization: `Bearer ${armToken}`,
        },
      });
      const planResult = planFetch.data;
      dispatch(updateCurrentPlan({ plan: planResult, loading: false }));
      return planResult;
    } catch (err) {
      console.log(err);
      dispatch(updateCurrentPlan({ loading: false }));
    }
    return null;
  };
}

export function updatePlan(value: ArmObj<ServerFarm>) {
  return async (dispatch: any, getState: () => IState) => {
    const startupInfo = getState().portalService.startupInfo;
    const armToken = startupInfo!.token;
    const armEndpoint = startupInfo!.armEndpoint;
    dispatch(updateCurrentPlanNoCache({ saving: true, updateFailed: false, updateFailedMessage: '' }));

    try {
      const planUpdate = await axios.put<ArmObj<ServerFarm>>(`${armEndpoint}${value.id}?api-version=2015-08-01`, value, {
        headers: {
          Authorization: `Bearer ${armToken}`,
        },
      });

      if (planUpdate.status !== 200) {
        dispatch(updateCurrentPlan({ saving: false, updateFailed: true, updateFailedMessage: planUpdate.statusText }));
      } else {
        const planResult = planUpdate.data;
        dispatch(updateCurrentPlan({ plan: planResult, saving: false, updateFailed: false, updateFailedMessage: '' }));
      }
    } catch (err) {
      console.log(err);
    }
  };
}
