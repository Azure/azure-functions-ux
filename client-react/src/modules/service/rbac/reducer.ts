import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import * as actions from './actions';
import {
  ADD_PERMISSION,
  ADD_PERMISSION_CALLED,
  ADD_READONLY_LOCK,
  ADD_READONLY_LOCK_CALLED,
  REMOVE_PERMISSION,
  REMOVE_PERMISSIONS_CALLED,
  REMOVE_READONLY_LOCK,
  REMOVE_READONLY_LOCK_CALLED,
} from './actionTypes';

export type RbacAction = ActionType<typeof actions>;
export interface IRbacState {
  permissionCalled: string[];
  readonlyLockCalled: string[];
  permissions: { [key: string]: boolean };
  readonlyLocks: { [key: string]: boolean };
}

export default combineReducers<IRbacState, RbacAction>({
  permissionCalled: (state = [], action) => {
    switch (action.type) {
      case ADD_PERMISSION_CALLED:
        return [...state, action.called];
      case REMOVE_PERMISSIONS_CALLED:
        return state.filter(x => x !== action.called);
      default:
        return state;
    }
  },
  readonlyLockCalled: (state = [], action) => {
    switch (action.type) {
      case ADD_READONLY_LOCK_CALLED:
        return [...state, action.called];
      case REMOVE_READONLY_LOCK_CALLED:
        return state.filter(x => x !== action.called);
      default:
        return state;
    }
  },
  permissions: (state = {}, action) => {
    switch (action.type) {
      case ADD_PERMISSION:
        return { ...state, [action.permissionKey]: action.value };
      case REMOVE_PERMISSION:
        const stateCopy = { ...state };
        delete stateCopy[action.permissionKey];
        return stateCopy;
      default:
        return state;
    }
  },
  readonlyLocks: (state = {}, action) => {
    switch (action.type) {
      case ADD_READONLY_LOCK:
        return { ...state, [action.resourceId]: action.lock };
      case REMOVE_READONLY_LOCK:
        const stateCopy = { ...state };
        delete stateCopy[action.resourceId];
        return stateCopy;
      default:
        return state;
    }
  },
});
