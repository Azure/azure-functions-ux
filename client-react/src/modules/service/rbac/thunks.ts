import { RootState, Services } from '../../../modules/types';
import {
  addPermission,
  addPermissionCalled,
  addReadonlyCalled,
  addReadonlyLock,
  removePermissionsCalled,
  removeReadonlyCalled,
} from './actions';

export interface PermissionCheckObj {
  resourceId: string;
  action: string;
}

export interface ReadonlyCheckObj {
  resourceId: string;
}

export function fetchPermissions(resources: PermissionCheckObj[]) {
  return async (dispatch: any, getState: () => RootState, { rbacHelper }: Services) => {
    await resources.map(async resource => {
      const resourceKey = `${resource.resourceId}|${resource.action}`;
      if (getState().rbac.permissionCalled.indexOf(resourceKey) === -1) {
        dispatch(addPermissionCalled(resourceKey));
        const permissionCheck = await rbacHelper.hasPermission(getState(), resource.resourceId, [resource.action!]);
        dispatch(addPermission({ permissionKey: resourceKey, value: permissionCheck }));
        dispatch(removePermissionsCalled(resourceKey));
      }
    });
  };
}

export function fetchReadonlyLocks(resources: ReadonlyCheckObj[]) {
  return async (dispatch: any, getState: () => RootState, { rbacHelper }: Services) => {
    await resources.map(async resource => {
      if (getState().rbac.readonlyLockCalled.indexOf(resource.resourceId) === -1) {
        dispatch(addReadonlyCalled(resource.resourceId));
        const permissionCheck = await rbacHelper.hasReadOnlyLock(resource.resourceId);
        dispatch(addReadonlyLock({ resourceId: resource.resourceId, lock: permissionCheck }));
        dispatch(removeReadonlyCalled(resource.resourceId));
      }
    });
  };
}
