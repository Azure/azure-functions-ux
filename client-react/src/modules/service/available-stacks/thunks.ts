import { updateAvailableStacks, updateStacksLoading } from './actions';
import { checkCacheValid } from 'redux-cache';
import axios from 'axios';
import { ArmArray } from '../../../models/WebAppModels';
import { AvailableStack } from '../../../models/available-stacks';

export function fetchStacks(osType = 'Windows') {
  return async (dispatch: any, getState: any) => {
    const startupInfo = getState().portalService.startupInfo;
    const armToken = startupInfo.token;
    const armEndpoint = startupInfo.armEndpoint;
    const currentStacksCall = getState().stacks;
    const isCacheValid = checkCacheValid(getState, 'stacks');
    if (isCacheValid || currentStacksCall.loading) {
      return;
    }

    dispatch(updateStacksLoading(true));

    try {
      const stacksFetch = await axios.get<ArmArray<AvailableStack>>(
        `${armEndpoint}/providers/Microsoft.Web/availableStacks?osTypeSelected=${osType}&api-version=2018-02-01`,
        {
          headers: {
            Authorization: `Bearer ${armToken}`,
          },
        }
      );
      const stacksResult = stacksFetch.data;
      dispatch(updateAvailableStacks(stacksResult));
    } catch (err) {
      console.log(err);
    } finally {
      dispatch(updateStacksLoading(false));
    }
  };
}
