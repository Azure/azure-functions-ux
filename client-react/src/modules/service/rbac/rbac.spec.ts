import {
  addPermission,
  addPermissionCalled,
  addReadonlyCalled,
  addReadonlyLock,
  removePermission,
  removePermissionsCalled,
  removeReadonlyCalled,
  removeReadonlyLock,
} from './actions';
import reducer from './reducer';
import { fetchPermissions, fetchReadonlyLocks } from './thunks';
import { Services } from '../../types';

const mockServices = (returnVal: boolean): Services =>
  ({
    rbacHelper: {
      hasPermission: async (resourceId: string, requestedActions: string[]): Promise<boolean> => {
        return returnVal;
      },
      hasReadOnlyLock: async (resourceId: string): Promise<boolean> => {
        return returnVal;
      },
    },
  } as Services);

describe('RBAC Store Reducer', () => {
  let initialState = reducer(undefined, {} as any);
  describe('initial state', () => {
    it('should match a snapshot', () => {
      expect(initialState).toMatchSnapshot();
    });

    it('loading should be false', () => {
      expect(initialState.permissionCalled.length).toBe(0);
      expect(initialState.readonlyLockCalled.length).toBe(0);
    });
  });

  describe('Rbac Fetch Stories', () => {
    it('Add new permission action', () => {
      const action = addPermission({ permissionKey: 'key', value: true });
      const state = reducer(initialState, action);
      expect(state.permissions['key']).toBe(true);
    });

    it('Remove permission state permission action', () => {
      let initialState = reducer(undefined, addPermission({ permissionKey: 'key', value: true }));
      expect(initialState.permissions['key']).toBeDefined();
      const action = removePermission('key');
      const state = reducer(initialState, action);
      expect(state.permissions['key']).toBeUndefined();
    });

    it('Add new permission called action', () => {
      const action = addPermissionCalled('key');
      const state = reducer(initialState, action);
      expect(state.permissionCalled).toContain('key');
    });

    it('Remove permission called action', () => {
      let initialState = reducer(undefined, addPermissionCalled('key'));
      expect(initialState.permissionCalled).toContain('key');
      const action = removePermissionsCalled('key');
      const state = reducer(initialState, action);
      expect(state.permissionCalled).not.toContain('key');
    });

    it('Add new readonly lock action', () => {
      const action = addReadonlyLock({ resourceId: 'resourceid', lock: true });
      const state = reducer(initialState, action);
      expect(state.readonlyLocks['resourceid']).toBe(true);
    });

    it('Remove read only lock action', () => {
      let initialState = reducer(undefined, addReadonlyLock({ resourceId: 'resourceid', lock: true }));
      expect(initialState.readonlyLocks['resourceid']).toBeDefined();
      const action = removeReadonlyLock('resourceid');
      const state = reducer(initialState, action);
      expect(state.readonlyLocks['resourceid']).toBeUndefined();
    });

    it('Add new readonly lock called action', () => {
      const action = addReadonlyCalled('resourceid');
      const state = reducer(initialState, action);
      expect(state.readonlyLockCalled).toContain('resourceid');
    });

    it('Remove read only lock called action', () => {
      let initialState = reducer(undefined, addReadonlyCalled('resourceid'));
      expect(initialState.readonlyLockCalled).toContain('resourceid');
      const action = removeReadonlyCalled('resourceid');
      const state = reducer(initialState, action);
      expect(state.permissionCalled).not.toContain('resourceid');
    });
  });
});

describe('RBAC Store thunks', () => {
  const dispatch = jest.fn();
  const getState = (calledId?: string) => () =>
    ({
      rbac: {
        permissionCalled: [calledId],
        readonlyLockCalled: [calledId],
      },
    } as any);
  afterEach(() => {
    dispatch.mockClear();
  });

  it('fetches permission', async () => {
    await fetchPermissions([
      {
        resourceId: 'resourceid',
        action: 'action',
      },
    ])(dispatch, getState(), mockServices(true));
    expect(dispatch.mock.calls).toEqual([
      [addPermissionCalled('resourceid|action')],
      [addPermission({ permissionKey: 'resourceid|action', value: true })],
      [removePermissionsCalled('resourceid|action')],
    ]);
  });

  it('does not fetch permissions again if a fetch is already in progress', async () => {
    await fetchPermissions([
      {
        resourceId: 'resourceid',
        action: 'action',
      },
    ])(dispatch, getState('resourceid|action'), mockServices(true));
    expect(dispatch).not.toHaveBeenCalled();
  });

  it('fetches readonly lock', async () => {
    await fetchReadonlyLocks([
      {
        resourceId: 'resourceid',
      },
    ])(dispatch, getState(), mockServices(true));
    expect(dispatch.mock.calls).toEqual([
      [addReadonlyCalled('resourceid')],
      [addReadonlyLock({ resourceId: 'resourceid', lock: true })],
      [removeReadonlyCalled('resourceid')],
    ]);
  });

  it('does not fetch readonly lock again if a fetch is already in progress', async () => {
    await fetchReadonlyLocks([
      {
        resourceId: 'resourceid',
      },
    ])(dispatch, getState('resourceid'), mockServices(true));
    expect(dispatch).not.toHaveBeenCalled();
  });
});
