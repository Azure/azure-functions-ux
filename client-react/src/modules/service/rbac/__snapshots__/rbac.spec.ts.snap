// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RBAC Store Reducer initial state should match a snapshot 1`] = `
Object {
  "permissionCalled": Array [],
  "permissions": Object {},
  "readonlyLockCalled": Array [],
  "readonlyLocks": Object {},
}
`;
